_posts/2020-02-24-qute.adoc

```diff
@@ -0,0 +1,198 @@
+</body>
+</html>
+----
+<1> Parameter declaration - maps `client` to `org.acme.Client`. See the <<type-safe-templates,next section>> for more information.
+<2> `org.acme.Client#getPullRequests()` is using a non-blocking Vert.x client to fetch the data directly from the GitHub API. Since the data resolution is asynchronous the thread is not blocked and can continue performing some other tasks:
++
+[source,java]
+----
+CompletionStage<JsonArray> getPullRequests() {
+   return webClient
+            .get(80, "api.github.com", "/repos/quarkusio/quarkus/pulls?state=open&per_page=10")
+            .as(BodyCodec.jsonArray())
+            .send()
+            .thenCompose(r -> {
+               if (r.statusCode() == 200) {
+                  return CompletableFuture.completedFuture(r.body());
+               } else {
+                  // Log errors etc.
+               }
+            });
+}
+----
+
+[[type-safe-templates]]
+=== Type-safe Templates
+
+Most of the templating engines are not type-safe, ie. do not prevent type errors.
+It's quite natural because dynamicity in templates is very often practical.
+On the other hand, a user is not protected from tedious errors caused by typos and various refactoring consequences.
+Qute templates can be _optionally_ type-safe.
+What does it actually mean?
+A template may contain one or more *parameter declarations*.
+A parameter declaration binds a concrete type information to a given identifier in the current context.
+And what are the advantages of having a type-safe template?
+
+* Quarkus validates _all expressions_ that reference a parameter declaration. If an invalid/incorrect expression is found the build fails. 
+
+TIP: In development mode, all files located in the `src/main/resources/templates` directory are watched for changes and modifications are immediately visible. That also implies that your application fails fast whenever a type error occurs.
+
+* A _value resolver_ is generated for all types used in parameter declarations so that it’s possible to access its properties without reflection. This is very useful when targeting GraalVM native images. 
+* We have few more ideas in our TODO list, such as performance optimizations for type-safe expressions, etc. 
+
+.Type-safe Template Example
+[source,html]
+----
+{@org.acme.Foo foo} <1>
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="UTF-8">
+<title>Qute Hello</title>
+</head>
+<body>
+  <h1>{title}</h1>  <2>
+  <p>{foo.message}</p>  <3>
+  {#for foo in baz.foos}
+    <p>Hello {foo.message}!</p>  <4>
+  {/for}
+</body>
+</html>
+----
+<1> Parameter declaration - maps `foo` to `org.acme.Foo`.
+<2> `{title}` is not validated - not matching a param declaration.
+<3> `{foo.message}` is validated. `org.acme.Foo` must have a property `message` or a matching template extension method must exist. 
+<4> `{foo.message}` is not validated because `foo` is overridden in the loop section and there is no type information available.
+
+=== First-class Quarkus Citizen
```

Comment:
this is a good point - I would though highlight Qute is usable on its own and not dependent on quarkus. Saw comments about users not wanting to use qute for templates as it "locked" them to quarkus...might as well remove that concern ;)